### PR TITLE
Leave common endpoint

### LINKED ISSUE #1221: Endpoint - Leave common
create endpoint for exit common

### PR DESCRIPTION
**Pull request by noamso1**

### CODE REVIEW

**Comment by noamso1** [comment]
no need to merge or review yet. still working on it.

File: functions/src/commons/services/leave.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ * @param userId - the current user
+ * @return true if success, error if error
+ */
+
+export const leave = async (payload: LeaveCommonPayload): Promise<boolean> => {
+    const { userId, commonId } = payload;
+    
+    // cancel all my subscriptions for this common's proposals
+    // ..this would have been so much easier if we had subscription.commonId
+    const commonProposals : IProposal[] = await proposalsDb.getMany( { commonId } )
```
Comment (SobolevYuri) [inline functions/src/commons/services/leave.ts:27]: I'd like to add a whitespace lines and semicolons to add the code more readable etc.
imho ;-)

File: functions/src/commons/services/leave.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    const commonProposalIds : string[] = commonProposals.map( e => e.id )
+    let mySubs : ISubscription[] = await subscriptionDb.getMany( { userId } )
+    mySubs = mySubs.filter( e => commonProposalIds.includes( e.proposalId ) )
+    for ( let sub of mySubs ) {
+       //sub.status = 'CanceledByUser'
```
Comment (SobolevYuri) [inline functions/src/commons/services/leave.ts]: you decided to clarify the further "cancelSubscription" functionality here, by adding the comments?

File: functions/src/commons/services/leave.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+export const leave = async (payload: LeaveCommonPayload): Promise<boolean> => {
+    const { userId, commonId } = payload;
+    
+    // cancel all my subscriptions for this common's proposals
+    // ..this would have been so much easier if we had subscription.commonId
+    const commonProposals : IProposal[] = await proposalsDb.getMany( { commonId } )
+    const commonProposalIds : string[] = commonProposals.map( e => e.id )
+    let mySubs : ISubscription[] = await subscriptionDb.getMany( { userId } )
+    mySubs = mySubs.filter( e => commonProposalIds.includes( e.proposalId ) )
+    for ( let sub of mySubs ) {
```
Comment (SobolevYuri) [inline functions/src/commons/services/leave.ts]: take a look at the warning, please
image

File: functions/src/commons/services/leave.ts (hunk 2 above)
Comment (SobolevYuri) [reply functions/src/commons/services/leave.ts]: if yes - I think that's no need

File: functions/src/commons/services/leave.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+export const leave = async (payload: LeaveCommonPayload): Promise<boolean> => {
+    const { userId, commonId } = payload;
+    
+    // cancel all my subscriptions for this common's proposals
+    // ..this would have been so much easier if we had subscription.commonId
+    const commonProposals : IProposal[] = await proposalsDb.getMany( { commonId } )
+    const commonProposalIds : string[] = commonProposals.map( e => e.id )
+    let mySubs : ISubscription[] = await subscriptionDb.getMany( { userId } )
```
Comment (SobolevYuri) [inline functions/src/commons/services/leave.ts]: mySubs -> userSubs/memberSubs maybe better?

a little bit weird user's appeal as if to himself

File: functions/src/commons/services/leave.ts (hunk 1 above)
Comment (noamso1) [reply functions/src/commons/services/leave.ts:27]: i don't understand exactly, can you show an example here ?

File: functions/src/commons/services/leave.ts (hunk 2 above)
Comment (noamso1) [reply functions/src/commons/services/leave.ts]: no, just before i implemented the cancelSubscription function, i did the update myself. no need now. removed.

File: functions/src/commons/services/leave.ts (hunk 3 above)
Comment (noamso1) [reply functions/src/commons/services/leave.ts]: Done!

File: functions/src/commons/services/leave.ts (hunk 4 above)
Comment (noamso1) [reply functions/src/commons/services/leave.ts]: Done. userSubs, userCommonSubs

File: functions/src/commons/services/leave.ts (hunk 1 above)
Comment (noamso1) [reply functions/src/commons/services/leave.ts:27]: added a whitespace line between the remark and the code..

**Comment by noamso1** [comment]
@SobolevYuri i'v done all of your remarks! please approve.

**Comment by SobolevYuri** [comment]
awesome, thanks!

so just refresh your branch with the latest dev, please

**Review by SobolevYuri — APPROVED** [review]

**Comment by ariel-dao** [comment]
@noamso1 need to make sure there are no active subscriptions for the user when he leaves common

**Comment by noamso1** [comment]
> @noamso1 need to make sure there are no active subscriptions for the user when he leaves common

Part of this dev was to cancel all your subscriptions before leaving the common.

**Pull request merged**